File: TypeScript/ShellSort.ts
// p106
const ShellSort = (datas: Array<number>) => {
    let span = datas.length / 2
    while (span > 0) {
        for (let i = 0; i < span; i++) {
            datas = InsertionSort(datas)
        }
        span /= 2
    }
    return datas
}

const InsertionSort = (datas: Array<number>) => {
    for (let i = 1; i < datas.length; i++) {
        let temp = datas[i]
        let j = i - 1
        while (j >= 0 && datas[j] > temp) {
            datas[j + 1] = datas[j]
            j -= 1
        }
        datas[j + 1] = temp
    }
    return datas
}

let datas: number[] = new Array
for (let i = 0; i < 10; i++) {
    datas.push(Math.floor(Math.random() * 100))
}

const sortedNums = ShellSort(datas.concat())

console.log('ソート前の要素の並び')
console.log(datas)
console.log('ソート後の要素の並び')
console.log(sortedNums)

File: Python/シェルソート/shell_sort.py
# p106


from typing import List


def shell_sort(datas: List[int]) -> List:

    span = len(datas)/2
    while span > 0:
        for _ in range(int(span)):
            datas = insertion_sort(datas)
        span /= 2
    return datas


def insertion_sort(datas: List[int]) -> List[int]:

    for i in range(1, len(datas)):
        temp = datas[i]
        j = i - 1
        while j >= 0 and datas[j] > temp:
            datas[j + 1] = datas[j]
            j -= 1
        datas[j + 1] = temp

    return datas


if __name__ == '__main__':
    from random import randint

    datas = [randint(0, 100) for _ in range(12)]

    sorted_datas = shell_sort(datas.copy())

    print("'ソート前の要素の並び'")
    print(datas)

    print("'ソート後の要素の並び'")
    print(sorted_datas)
